pulsar-broker/src/test/java/org/apache/pulsar/client/impl/KeySharedSubscriptionTest.java

```diff
@@ -0,0 +1,170 @@
+                        try {
+                            cons2.acknowledge(msg);
+                        } catch (PulsarClientException e) {
+                            throw new RuntimeException(e);
+                        }
+                    }
+                })
+                .consumerName("con-2")
+                .subscribe();
+        @Cleanup
+        Consumer<byte[]> consumer3 = pulsarClient.newConsumer()
+                .topic(topic)
+                .subscriptionName("sub-1")
+                .subscriptionType(SubscriptionType.Key_Shared)
+                .messageListener((cons3, msg) -> {
+                    lastActiveTime.set(System.currentTimeMillis());
+                    nameToId.computeIfAbsent(cons3,(k) -> new ArrayList<>())
+                            .add(msg.getMessageId());
+                    msgCount.incrementAndGet();
+                    if (canAcknowledgement.get()) {
+                        try {
+                            cons3.acknowledge(msg);
+                        } catch (PulsarClientException e) {
+                            throw new RuntimeException(e);
+                        }
+                    }
+                })
+                .consumerName("con-3")
+                .subscribe();
+
+        @Cleanup
+        Producer<byte[]> producer = pulsarClient.newProducer()
+                .topic(topic)
+                .enableBatching(true)
+                .create();
+
+        for (int i = 0; i < 50000; i++) {
+            producer.sendAsync(UUID.randomUUID().toString().getBytes(StandardCharsets.UTF_8));
+        }
+
+        // Wait for all consumers can not read more messages. the consumers are stuck by max unacked messages.
+        while (true) {
+            long differentTime = System.currentTimeMillis() - lastActiveTime.get();
+            if (differentTime > TimeUnit.SECONDS.toMillis(20)) {
+                break;
+            }
+        }
+
+        // All consumers can acknowledge messages as they continue to receive messages.
+        canAcknowledgement.set(true);
+
+        // Acknowledgment of currently received messages to get out of stuck state due to unack message
+        for (Map.Entry<Consumer<?>, List<MessageId>> entry : nameToId.entrySet()) {
+            Consumer<?> consumer = entry.getKey();
+            consumer.acknowledge(entry.getValue());
+        }
+        // refresh active time
+        lastActiveTime.set(System.currentTimeMillis());
+
+        // Wait for all consumers to continue receiving messages.
+        while (true) {
+            long differentTime = System.currentTimeMillis() - lastActiveTime.get();
+            if (differentTime > TimeUnit.SECONDS.toMillis(20)) {
+                break;
+            }
+        }
+
+        //Determine if all messages have been received.
+        //If the dispatcher is stuck, we can not receive enough messages.
+        Assert.assertEquals(msgCount.get(), 50000);
```

Comment:
Maybe we need to check the messages not only the messages count, for example if we missed 2 messages, but 2 messages is consumed duplicated, the test will also get passed.

Reply:
fixed